src/pykiso/tool/testrail/api.py

```diff
@@ -0,0 +1,394 @@
+    API
+
+.. currentmodule:: api
+
+"""
+
+import base64
+import concurrent.futures
+import json
+from enum import Enum
+from typing import Callable, List
+
+import requests
+
+
+class ContentType(Enum):
+    """Store all available header's content-type."""
+
+    #: JSON content type
+    JSON: str = "application/json"
+
+
+class ApiVersion(Enum):
+    """Store all available TestRail rest API version."""
+
+    #: TestRail's V2 API
+    V2: str = "index.php?/api/v2/"
+
+
+class HttpError(Exception):
+    """Raise if the status code from a HTTP request is different than
+    200.
+    """
+
+    def __init__(self, status_code: str, reason: str) -> None:
+        """Initialize error message attributes.
+
+        :param status_code: response status code
+        :param reason: response content
+        """
+        self.status_code = status_code
+        self.reason = reason
+
+    def __str__(self) -> str:
+        """Exception message representation."""
+        return f"HTTP request exit with status {self.status_code} : {self.reason}"
+
+
+def handle_http_error(func: Callable) -> Callable:
+    """Decorator used to handle HTTP request negative response
+
+    :param func: inner handle function
+
+    :return: the function returned value
+
+    :raises Exception: if the response status code is different than
+            200
+    """
+
+    def handle(*args: tuple, **kwargs: dict) -> requests.Response:
+        """Handle sttaus code different than 200.
+
+        :param args: positonnal arguments
+
+        :return: kwargs named arguments
+
+        :raises Exception: if the response status code is different than
+                200
+        """
+        response = func(*args, **kwargs)
+
+        if response.status_code != 200:
+            raise HttpError(response.status_code, response.content)
+
+        return response
+
+    return handle
+
+
+class HttpRequest:
+    """Encapsulate all available HTTP requests."""
+
+    @staticmethod
+    @handle_http_error
+    def get(url: str, headers: dict) -> requests.Response:
+        """Send a GET HTTP request and return the received response.
+
+        :param url: request target url
+        :param headers: request's header (authorization, content-type)
+
+        :return: received response from server
+
+        :raises HttpError: if the response status code is different than
+            200
+        """
+        return requests.get(url, headers=headers)
+
+    @staticmethod
+    @handle_http_error
+    def post(url: str, headers: dict, data: dict) -> requests.Response:
+        """Send a POST HTTP request and return the received response.
+
+        :param url: request target url
+        :param headers: request's header
+        :param data: reauest body
+
+        :return: received response from server
+
+        :raises HttpError: if the response status code is different than 200
+        """
+        return requests.post(url, headers=headers, data=data)
+
+
+class TestRailApi:
```

Comment:
In a quick overview, I do not find it, is there a check to know if it is a HTTPS request? I don't think we should send a password in base64 if it is not https...

Reply:
We are the Client side so it will use the protocol in place.

Reply:
In addition if I use the verify flag the cli utility is unusable for self-signed certificate (instead of True path to user's self-signed certificate)

Reply:
We should ensure that we do not send the password unsecured.

Reply:
I did not get your second comment

Reply:
If somebody is using a self-signed certificate the request will be:
requests.post(url, data=data, verify='/path/to/certificate')

Otherwise the following post request will be:
requests.post(url, data=data, verify=True)

This means from TestRail CLI utility to integrate a new optional argument certificate in order to pass it from cli to api.

Reply:
At least now the verify flag is up to True